### PR TITLE
❗ [Includes historical git commits] Merge latest version of ministryofjustice-data-platform-catalogue

### PR DESCRIPTION
**Pull request by MatMoore**
This PR supersedes https://github.com/ministryofjustice/find-moj-data/pull/300.

There is a corresponding PR to remove this from the old repo: https://github.com/ministryofjustice/data-platform/pull/4261

## Merge library into this repo

This library was previously in https://github.com/ministryofjustice/data-platform/
I've used `git filter-repo` to preserve the git history, so this PR contains the entire history of this library.

For reference, the command I ran on the source repo was

```
git filter-repo --path python-libraries/data-platform-catalogue --path-rename python-libraries/data-platform-catalogue:lib/datahub-client
```

Which allowed me to pull in the commits with 

```
git pull --allow-unrelated-histories ../dp-clone main --no-ff
```

## Post-refactor fixes
Since we made breaking changes in the 1.0.0 refactor, this PR also contains @murdo-moj's work to update the django project.

## What's still missing
CI runs both the django tests and the tests for lib/datahub-client. But we will need to push another workflow to publish the library so it can be used by our metadata ingestion code.

### CODE REVIEW

**Comment by MatMoore** [comment]
Note: We will need to switch off squash and merge to preserve this history

**Comment by MatMoore** [comment]
This has not yet fixed the `String should match pattern '^[\w.]+$' [type=string_pattern_mismatch, input_value='[version=2.0].[type=int].year', input_type=str]
    For further information visit https://errors.pydantic.dev/2.7/v/string_pattern_mismatch` bug

- I'll follow up on this separately

**Comment by MatMoore** [comment]
Need to make sure the library is installed properly in prod

Currently it is a dev dependency and we do this

`RUN poetry install --without dev --no-root && rm -rf $POETRY_CACHE_DIR`

File: .vscode/launch.json [hunk 1]
```diff
@@ -10,6 +10,7 @@
       "request": "launch",
       "program": "${workspaceFolder}/manage.py",
       "args": ["runserver"],
+      "justMyCode": false,
```
Comment (murdo-moj) [inline .vscode/launch.json:13]:
```suggestion
      // "justMyCode": false,
```

File: templates/details_table.html [hunk 2]
```diff
@@ -69,7 +69,7 @@ <h3 class="govuk-heading-s govuk-!-margin-top-3">
       </div>
     </div>
     <div class="govuk-grid-column-one-third">
-      {% include "partial/contact_info.html" with data_owner=table.owner data_owner_email=table.owner_email %}
+      {% include "partial/contact_info.html" with data_owner=table..governance.data_owner.display_name data_owner_email=table..governance.data_owner.email %}
```
Comment (murdo-moj) [inline templates/details_table.html:72]:
```suggestion
      {% include "partial/contact_info.html" with data_owner=table.governance.data_owner.display_name data_owner_email=table.governance.data_owner.email %}
```

File: templates/partial/search_result.html [hunk 3]
```diff
@@ -38,43 +34,23 @@ <h3 class="govuk-heading-m govuk-!-margin-bottom-2">
...
-            <dd class="govuk-summary-list__value">
-              {{ result.tags |join:", " }}
-            </dd>
-          </div>
           {% if result.matches %}
           <div class="govuk-summary-list__row">
             <dt class="govuk-summary-list__key">Matched fields</dt>
             <dd class="govuk-summary-list__value">
-              {{ result.matches|lookup:readable_match_reasons|join:", " }}
+              {{ result.matches|join:", " }}
```
Comment (murdo-moj) [inline templates/partial/search_result.html:53]:
```suggestion
              {{ result.matches|lookup:readable_match_reasons|join:", " }}
```
The match reasons dict is missing some items, this pulls through a blank sometimes. Easy to add to.

File: tests/conftest.py [hunk 4]
```diff
@@ -1,26 +1,23 @@
+from datetime import datetime, timezone
 from random import choice
 from typing import Any
 from unittest.mock import MagicMock, patch
 
 import pytest
-from data_platform_catalogue.client import BaseCatalogueClient
-from data_platform_catalogue.entities import RelationshipType, TableMetadata
+from data_platform_catalogue.client.datahub_client import DataHubCatalogueClient
+from data_platform_catalogue.entities import RelationshipType, Table, DomainRef, Governance, OwnerRef, TagRef, Column, ColumnRef, UsageRestrictions, EntityRef, CustomEntityProperties, DataSummary, AccessInformation
```
Comment (murdo-moj) [inline tests/conftest.py:8]: The length of this line makes me think we could run black/isort on the codebase

**Review by tom-webber — APPROVED** [review]

**Review by hjribeiro-moj — APPROVED** [review]
lgtm

**Pull request merged**